drafts/keep-the-lights-on.md

```diff
@@ -0,0 +1,64 @@
+
+## Keep the lights on
+
+I have been talking about the root cause of missing proper error handling. The question remains, what can we do to prevent these error dialogs? There are some cases when nothing works anymore but in many cases, we handle them and keep the application running.
+
+It is better to degrade the functionality of an application than let it stop working. It is important to inform the user about the misbehavior, but let him keep doing his work within the limits of what is still possible. This way the user can make its own decision on how to proceed.
+
+Let's give some examples:
+
+We display personal information in our app and resolve the person's address from an external address service. When this service returns a failure or is unreachable we can display that information to the user.
+
+![Degraded address information example](https://cdn.hashnode.com/res/hashnode/image/upload/v1709018312932/mDYMElQpY.png?auto=format)
+
+This way, the user knows that the system currently has a problem, but it is possible to continue working as long as no functionality using the address service needs to be used. Maybe it can be one later when it is working again.
+
+Another example could be a feature where several checks get executed. Let's say our software checks 'conflicts of interests' between multiple persons. It checks different constraints such as whether the persons are working in the same institution or have worked on a shared project etc. The data gets collected from different APIs and each check result gets displayed so that the user can decide whether a person is suitable or not. Now let's assume one of the services providing check-related data is currently not working. We can encapsulate this failure and still show the check results of all working services and we inform the user that a certain check could not be executed due to some failures.
+
+![System failure in checks example](https://cdn.hashnode.com/res/hashnode/image/upload/v1711525860862/Gt9UnHmOy.png?auto=format)
+
+Again, this way the user can decide whether it's safe to still go on with this person or not.
+
+There are many ways how we can handle errors with feature degradation. It is not always easy to find a way, but I would argue that in most cases it is possible! Just be creative and involve people.
+
+To end the article I would like to recommend you one of my favorite books about failure handling: [Release It!](https://www.amazon.de/Release-Production-Ready-Software-Pragmatic-Programmers/dp/0978739213?language=en_GB&currency=EUR)
+
+## Summary
+
+System downtime can cost a fortune or have legal impacts. But usually, we are focused on feature delivery. This ends with an error popup developed by the software engineers. Applications must be designed for errors as well, so we must include the whole team in that process. It is better to handle errors and degrade some functionalities instead of letting the application stop working.
```

Comment:
```suggestion
System downtime can cost a fortune or have legal impacts. But usually, we are focused on feature delivery. Leading to a general error popup developed by the software engineers. Applications must be designed for errors as well, so we must include the whole team in that process. It is better to handle errors and degrade some functionalities instead of letting the application stop working.
```